### ISSUE TITLE
Can jackson go faster?

### ISSUE THREAD

**Issue by fabienrenaud** [FasterXML/jackson-core#379]
I am the author of https://github.com/fabienrenaud/java-json-benchmark which compares simple serialization and deserialization throughput performance of a variety of json libs, including jackson.
I recently added dsl-json and jsoniter to the benchmarks and was surprised to see how much faster than jackson these are on some metrics.

The author of jsoniter wrote an article that dives a little into what dsl-json/jsoniter are optimized against and how they do it: https://dzone.com/articles/is-protobuf-5x-faster-than-json and https://dzone.com/articles/is-protobuf-5x-faster-than-json-part-ii

Can some of these performance boosters be applied to jackson?

**Comment by cowtowncoder** [comment]
Perhaps. But one concern I have is that these focus heavily on number handling, which really is not at all what most time is usually spent. I would almost say that these are benchmarks created to bolster article, instead of focusing on common use case. Or perhaps just to emphasize where protobuf shines -- not a big surprise, IEEE `double`/`float` handling.
I mean there may well be use case where arrays of numbers is the dominant format; and for these cases I assume improvement can be made. But I am not sure how important that is.
I have similar concerns regarding input size: there is certain amount of fixed overhead per parsing call, and this is more evident for tiny documents. However, test uses 10 numbers so this might be ok.

Another question I have is that what many newer parser tend to skim on are things like:

1. Error detection and handling: often detection of invalid input is lacking; and in cases where it is detected, location information is missing or inaccurate
2. Support for non-standard features that many users require: comments especially (but occasionally other features)

The reason I mention these is that if one ignores both aspects, it is possible to apply some optimizations that are not available otherwise. At one point, for example, dsljava omitted actual verification of field names, fully trusting hash codes calculated from names. This can lead to potential (although unlikely, obviously) false matches. Similarly for number decoding it is possible to skip certain checks.

Anyway: one thing that could be helpful would be to have micro-benchmarks to use. Perhaps ones included could help. And although number handling is actually somewhat optimized already, it might be possible to add some tweaks.

**Comment by cowtowncoder** [comment]
Note: above notes are mostly wrt article. I did look at `dsljava`, its optimizations earlier, and it is quite impressive indeed. Unfortunately improvements to, say, field name handling seemed difficult to replicate: while extremely clever, it is quite tied to the way input is handled.
Another lib that showed impressive performance in some other tests was `fastjson`; some learnings from it were included in afterburner.

One question on `java-json-benchmark`: what kind of input data does it use/generate? I think I had a look at some point, but I can't remember details now. I don't think it focused heavily on numbers.

**Comment by fabienrenaud** [comment]
> One question on java-json-benchmark: what kind of input data does it use/generate?

The default schema is very similar to the one http://www.json-generator.com/ uses and its POJO is here: [Users.java](https://github.com/fabienrenaud/java-json-benchmark/blob/master/src/main/java/com/github/fabienrenaud/jjb/model/Users.java) (so mostly strings). The generator is here: [UsersGenerator.java](https://github.com/fabienrenaud/java-json-benchmark/blob/8565922a8366f78253747adbee3f371e74a18f18/src/main/java/com/github/fabienrenaud/jjb/data/gen/UsersGenerator.java)

> But one concern I have is that these focus heavily on number handling

That is the first part of the article, yes. [The second part](https://dzone.com/articles/is-protobuf-5x-faster-than-json-part-ii) focuses on object binding and strings. And the author also mentions these tests are extremely biased towards 1 measuring one thing at the type rather than 1 mix of everything (like my benchmark does).
However, relative performance comparisons between jackson, dsl-json and jsoniter always puts jackson last. So, indeed it might be lack of quality error handling or feature support but some of the simple optimizations for reading ints and ascii strings in combination with javassist may be interesting to pull in.

> dsljava omitted actual verification of field names, fully trusting hash codes calculated from names. This can lead to potential (although unlikely, obviously) false matches.

I agree. even in strict mode it goes faster than jackson.

> one thing that could be helpful would be to have micro-benchmarks to use.

That'd be nice yeah.



Overall, I am not saying jackson does it wrong but rather there may be some optimizations in dsl-json or jsoniter or other that may be worth looking at and integrating in jackson. :)
Maybe with config levers/switches for some (for example: if know 99.99% of my data are pure ascii strings, i'd like to use an ascii optimization)...

**Comment by cowtowncoder** [comment]
Right, I appreciate all ideas for improving performance, and did not take this as criticism at all.
I am always open to improvements.
Thank you also for link to second part; I had missed it indeed.

It is interesting to see that many of optimizations mentioned are used in some form by Jackson too.
I think much overhead Jackson has is probably more related to complexity of error handling, as well as (more) modular design (which leads to need to retain more state). One example is that decoding of Strings by jsoniter did not check for invalid content (unescaped linefeeds); this can be reasonable (only works with fully valid json), but can be problematic in accepting invalid content.
None of these checks alone explains the difference -- and difference is surprisingly high, dsl-java/jsoniter are impressively optimized -- but they add up.

One thing that would be nice would be to figure out good ways for more micro-benchmarking.
I have relied on old `jvm-serializers` a lot; but it is definitely skewed towards certain usage.
It is sometimes good to have more biased ones, to run multiple ones when trying out new optimization strategies.

**Comment by hc-codersatlas** [comment]
@fabienrenaud when's the next run? There's quite a number of ways Jackson can be optimized from what I see - mostly from newer JVM changes. Happy to help see where I can contribute and make Jackson top of the graph.

**Comment by fabienrenaud** [comment]
I usually do full re-runs when there are significant changes in perf
results from the top libs or when a new lib is added and performs well.
The last run was in April. Without the significant changes I just
mentioned, the normal cadence is once a year...
But you can always run the project yourself on your laptop :)

Also, the project is in Java 8 at the moment. Adoption of java 9/10 will be
slow, so it’ll stay in 8 for a while...

**Comment by cowtowncoder** [comment]
Fwtw, here are some links that could help esp. with last 2 types:

https://www.reddit.com/r/java/comments/darehu/jackson_release_210/f1ysf1e/
